src/test/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/resource/validator/UpdateGoalRequestReferenceConstraintValidatorTest.kt

```diff
@@ -21,8 +22,11 @@ class UpdateGoalRequestReferenceConstraintValidatorTest {
     UpdateGoalRequest::class.java,
     String::class.java,
     UUID::class.java,
+    Continuation::class.java,
```

Comment:
This is the test that I mentioned above re: the use of the `suspend` function affecting the java byte code.
In this case we are using reflection to get the method from the controller. As part of the reflection call we need to specify the method name and it's arguments, then reflection does it's magic and finds the method.

Because we have marked the method as a `suspend` function in the controller, it has changed the method in the byte code by adding a new arg to it - the `Continuation` argument. Note - this does not change the argument list in the source code or that calling code need to call it with; it's just at the byte code that it's different.

Sad actual times 😢

Reply:
Sad times indeed! 😆 I can see why you want unit tests here but is the problem that Kotlin suspend functions are dirty, or is the problem that using reflection in this way is less than ideal? I'd be tempted to try and achieve the same with an integration test. For me that would give more confidence that your constraints are working as intended, this feels a little bit contrived?

Reply:
I take your points, though I'm not sure the test is that contrived.
We have a couple of custom constraint validators and generally our view re testing them is to test them 'thoroughly' in unit tests, and test them once in an integration test. Unit tests are cheap and fast; integration tests less so
For example we have a constraint validator that asserts a date is not in the past. Our unit test tests the validator with "today", "tomorrow" and "yesterday" (and it would 'cost' nothing to add more dates/scenarios if the complexity of the validator warranted it).
We also have this date validator covered in one integration test. That test is not about testing whether the validator works in all scenarios, but is about essentially testing whether everything is wired up and that the request is a rejected with a 400 with the correct messages etc.

Back to this validator and why it has a unit test .... whilst the validator itself is simple (does the reference number in the request URI path parameter equals the reference number in the request body object?) the implementation is quite complex in that it needs to be a `SupportedValidationTarget` validator in order to get the controller method args rather than an annotation on a single field. The test needs a reference to the controller class and the method that the annotation is used on; hence the nastiness of reflection in order to get the `Method` object.

Reply:
All good points. Though I would say that the old testing pyramid is a bit of a different shape these days... more of a trapezoid? 😆 Since Docker came along, integration tests aren't half as expensive as they used to be (though it is a bit painful when you want to run a single integration test locally and you have to wait for Spring to start up). For example we've given up on unit tests where we're interested in the database, integration tests only, mocking repositories is painful and of little value.